### ISSUE TITLE
False positive for only_used_in_recursion for trait implementations

### ISSUE THREAD

**Issue by DerGut** [rust-lang/rust-clippy#14566]
### Summary

I stumbled over this warning in a trait implementation. The issue is that a trait defines a method with multiple parameters, but not all its implementations use all the parameters. Some will simply pass the parameter to recursive calls. However, they need to still define (and pass) the parameter to conform to the trait definition.

More concretely, I identified this issue with the [`physical_optimizer::AggregateStatistics::optimize`](https://github.com/apache/datafusion/blob/5a6db9f5d4bb714b820c1b960ff1ed2ce897c872/datafusion/physical-optimizer/src/aggregate_statistics.rs#L45) implementation of the [`physical_optimizer::PhysicalOptimizerRule::optimize`](https://github.com/apache/datafusion/blob/5a6db9f5d4bb714b820c1b960ff1ed2ce897c872/datafusion/physical-optimizer/src/optimizer.rs#L50) trait in the DataFusion project.

In my opinion, Clippy needs to take more context into account when the warning is triggered in a trait implementation. For example, when the rule triggers for all implementations of a trait (and it is private), then a warning could still suggest to remove that parameter. If at least one implementation is using the parameter, it shouldn't trigger.
A more incremental improvement would be to simply not trigger whenever this warning is triggered on a trait implementation.

---

I tested this with Rust `v1.85.0` and `v1.86.0`.

### Lint Name

only_used_in_recursion

### Reproducer

I tried this code:

```rust
trait DoSomething {
    fn do_something(&self, a: i32, b: i32) -> i32;
}

struct Impl1 {}

// This works fine
impl DoSomething for Impl1 {
    fn do_something(&self, a: i32, b: i32) -> i32 {
        a + b
    }
}


struct Impl2 {}

// I get a a warning here
impl DoSomething for Impl2 {
    fn do_something(&self, a: i32, b: i32) -> i32 {
        if a == 0 {
            self.do_something(a, b)
        } else {
            self.do_something(a - 1, b)
        }
    }
}
```

I saw this happen:

```
warning: parameter is only used in recursion
  --> src/test.rs:24:36
   |
24 |     fn do_something(&self, a: i32, b: i32) -> i32 {
   |                                    ^ help: if this is intentional, prefix it with an underscore: `_b`
   |
note: parameter used here
  --> src/test.rs:26:34
   |
26 |             self.do_something(a, b)
   |                                  ^
27 |         } else {
28 |             self.do_something(a - 1, b)
   |                                      ^
```

I expected to see this happen:

I expect it to pass.

### Version

```text

```

**Comment by relaxcn** [comment]
I don’t think this should be considered a false positive. A static analysis lint should not judge based on subjective intent, but rather on objective facts.

Even if, in your code, you are forced to use a useless variable, from the code's perspective alone, it is indeed problematic (as it contains an useless variable).

That said, your suggestion does make sense (to check all trait implementations), but as things stand, this would require analyzing the entire crate from start and storing intermediate results in memory, which, at least to me, seems unreasonable.

Until a better solution is found, I think it would be more appropriate to disable this lint in your code.

In addition, I have some doubts, why not all a trait's implementations use all the parameters? I think your good design should not allow this to happen.

But this is just my personal opinion, I'm a newbie in rust :)

**Comment by relaxcn** [comment]
Do you have any insights on this issue? @samueltardieu

**Comment by samueltardieu** [comment]
> Do you have any insights on this issue? [@samueltardieu](https://github.com/samueltardieu)

Please avoid tagging me in discussions, unless *me specifically* not participating blocks things.

**Comment by relaxcn** [comment]
I apologize for the inconvenience.

> > Do you have any insights on this issue? [@samueltardieu](https://github.com/samueltardieu)
> 
> Please avoid tagging me in discussions, unless _me specifically_ not participating blocks things.
